File: src/main/java/me/saro/kit/Dates.java
package me.saro.kit;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;

/**
 * date util
 * @author		PARK Yong Seo
 * @since		1.0.0
 */
public class Dates {

    public static Map<String, DateTimeFormatter> formatters = new HashMap<>();
    public static Map<String, SimpleDateFormat> simpleDateFormats = new HashMap<>();

    public static ZonedDateTime toZonedDateTime(Calendar calendar) {
        var tz = calendar.getTimeZone();
        return ZonedDateTime.ofInstant(calendar.toInstant(), tz != null ? tz.toZoneId() : ZoneId.systemDefault());
    }

    public static LocalDateTime toLocalDateTime(Calendar calendar) {
        var tz = calendar.getTimeZone();
        return LocalDateTime.ofInstant(calendar.toInstant(), tz != null ? tz.toZoneId() : ZoneId.systemDefault());
    }

    public static LocalDate toLocalDate(Calendar calendar) {
        var tz = calendar.getTimeZone();
        return LocalDate.ofInstant(calendar.toInstant(), tz != null ? tz.toZoneId() : ZoneId.systemDefault());
    }

    public static LocalTime toLocalTime(Calendar calendar) {
        var tz = calendar.getTimeZone();
        return LocalTime.ofInstant(calendar.toInstant(), tz != null ? tz.toZoneId() : ZoneId.systemDefault());
    }

    public static ZonedDateTime parseZonedDateTime(String date, String format, ZoneId zone) {
        return ZonedDateTime.parse(date, toFormatter(format).withZone(zone));
    }

    public static ZonedDateTime parseZonedDateTime(String date, String format) {
        return ZonedDateTime.parse(date, toFormatter(format));
    }

    public static LocalDateTime parseLocalDateTime(String date, String format) {
        return LocalDateTime.parse(date, toFormatter(format));
    }

    public static LocalDate parseLocalDate(String date, String format) {
        return LocalDate.parse(date, toFormatter(format));
    }

    public static LocalTime parseLocalTime(String date, String format) {
        return LocalTime.parse(date, toFormatter(format));
    }

    public static Calendar parseCalendar(String date, String format) {
        var calendar = Calendar.getInstance();
        calendar.setTime(parseDate(date, format));
        return calendar;
    }

    public static Date parseDate(String date, String format) {
        try {
            return toSimpleDateFormat(format).parse(date);
        } catch (ParseException e) {
            throw new DateTimeParseException(e.getMessage(), date, -1, e);
        }
    }

    public static synchronized DateTimeFormatter toFormatter(String format) {
        DateTimeFormatter formatter;
        if ((formatter = formatters.get(format)) == null) {
            formatters.put(format, (formatter = DateTimeFormatter.ofPattern(format)));
        }
        return formatter;
    }

    public static synchronized SimpleDateFormat toSimpleDateFormat(String format) {
        SimpleDateFormat simpleDateFormat;
        if ((simpleDateFormat = simpleDateFormats.get(format)) == null) {
            simpleDateFormats.put(format, (simpleDateFormat = new SimpleDateFormat(format)));
        }
        // SimpleDateFormat is not thread-safe
        // solution is return cloned object
        return (SimpleDateFormat)simpleDateFormat.clone();
    }

    public static Calendar toCalendar(Date date) {
        var calendar = Calendar.getInstance();
        calendar.setTime(date);
        return calendar;
    }

    public static Calendar toCalendar(ZonedDateTime zonedDateTime) {
        var calendar = Calendar.getInstance();
        calendar.clear();
        calendar.set(
                zonedDateTime.getYear(), zonedDateTime.getMonthValue() - 1, zonedDateTime.getDayOfMonth(),
                zonedDateTime.getHour(), zonedDateTime.getMinute(), zonedDateTime.getSecond()
        );
        calendar.set(Calendar.MILLISECOND, Math.round(zonedDateTime.getNano() / 1_000_000));
        calendar.setTimeZone(TimeZone.getTimeZone(zonedDateTime.getZone()));
        return calendar;
    }

    public static Calendar toCalendar(LocalDateTime localDateTime) {
        var calendar = Calendar.getInstance();
        calendar.clear();
        calendar.set(
                localDateTime.getYear(), localDateTime.getMonthValue() - 1, localDateTime.getDayOfMonth(),
                localDateTime.getHour(), localDateTime.getMinute(), localDateTime.getSecond()
        );
        calendar.set(Calendar.MILLISECOND, localDateTime.getNano() / 1_000_000);
        return calendar;
    }

    public static Calendar toCalendar(LocalDate localDate) {
        var calendar = Calendar.getInstance();
        calendar.clear();
        calendar.set(localDate.getYear(), localDate.getMonthValue() - 1, localDate.getDayOfMonth());
        return calendar;
    }

    public static Calendar toCalendar(LocalTime localTime) {
        var calendar = Calendar.getInstance();
        calendar.clear();
        calendar.set(Calendar.HOUR, localTime.getHour());
        calendar.set(Calendar.MINUTE, localTime.getMinute());
        calendar.set(Calendar.SECOND, localTime.getSecond());
        calendar.set(Calendar.MILLISECOND, localTime.getNano() / 1_000_000);
        return calendar;
    }
}

File: src/main/kotlin/me/saro/kit/Dates.kt
package me.saro.kit

import java.text.ParseException
import java.text.SimpleDateFormat
import java.time.*
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.*

/**
 * date util
 * @author PARK Yong Seo
 * @since 1.0.0
 */
open class Dates {

    companion object {
        private val formatters: MutableMap<String, DateTimeFormatter> = HashMap()
        private val simpleDateFormats: MutableMap<String, SimpleDateFormat> = HashMap()

        /**
         *
         * @param calendar
         * @return
         */
        @JvmStatic
        fun toZonedDateTime(calendar: Calendar): ZonedDateTime {
            val tz = calendar.timeZone
            return ZonedDateTime.ofInstant(calendar.toInstant(), if (tz != null) tz.toZoneId() else ZoneId.systemDefault())
        }

        /**
         *
         * @param calendar
         * @return
         */
        @JvmStatic
        fun toLocalDateTime(calendar: Calendar): LocalDateTime {
            val tz = calendar.timeZone
            return LocalDateTime.ofInstant(calendar.toInstant(), if (tz != null) tz.toZoneId() else ZoneId.systemDefault())
        }

        /**
         *
         * @param calendar
         * @return
         */
        @JvmStatic
        fun toLocalDate(calendar: Calendar): LocalDate {
            val tz = calendar.timeZone
            return LocalDateTime.ofInstant(calendar.toInstant(), if (tz != null) tz.toZoneId() else ZoneId.systemDefault())
                .toLocalDate()
        }

        /**
         *
         * @param calendar
         * @return
         */
        @JvmStatic
        fun toLocalTime(calendar: Calendar): LocalTime {
            val tz = calendar.timeZone
            return LocalDateTime.ofInstant(calendar.toInstant(), if (tz != null) tz.toZoneId() else ZoneId.systemDefault())
                .toLocalTime()
        }

        /**
         *
         * @param date
         * @param format
         * @param zone
         * @return
         */
        @JvmStatic
        fun parseZonedDateTime(date: String?, format: String, zone: ZoneId?): ZonedDateTime {
            return ZonedDateTime.parse(date, toFormatter(format).withZone(zone))
        }

        /**
         *
         * @param date
         * @param format
         * @return
         */
        @JvmStatic
        fun parseZonedDateTime(date: String?, format: String): ZonedDateTime {
            return ZonedDateTime.parse(date, toFormatter(format))
        }

        /**
         *
         * @param date
         * @param format
         * @return
         */
        @JvmStatic
        fun parseLocalDateTime(date: String?, format: String): LocalDateTime {
            return LocalDateTime.parse(date, toFormatter(format))
        }

        /**
         *
         * @param date
         * @param format
         * @return
         */
        @JvmStatic
        fun parseLocalDate(date: String?, format: String): LocalDate {
            return LocalDate.parse(date, toFormatter(format))
        }

        /**
         *
         * @param date
         * @param format
         * @return
         */
        @JvmStatic
        fun parseLocalTime(date: String?, format: String): LocalTime {
            return LocalTime.parse(date, toFormatter(format))
        }

        /**
         *
         * @param date
         * @param format
         * @return
         */
        @JvmStatic
        fun parseCalendar(date: String?, format: String): Calendar {
            val calendar = Calendar.getInstance()
            calendar.time = parseDate(date, format)
            return calendar
        }

        /**
         *
         * @param date
         * @param format
         * @return
         */
        @JvmStatic
        fun parseDate(date: String?, format: String): Date {
            return try {
                toSimpleDateFormat(format).parse(date)
            } catch (e: ParseException) {
                throw DateTimeParseException(e.message, date, -1, e)
            }
        }

        /**
         *
         * @param format
         * @return
         */
        @Synchronized
        @JvmStatic
        fun toFormatter(format: String): DateTimeFormatter {
            var formatter: DateTimeFormatter
            if (formatters[format].also { formatter = it!! } == null) {
                formatters[format] = DateTimeFormatter.ofPattern(format).also { formatter = it }
            }
            return formatter
        }

        /**
         *
         * @param format
         * @return
         */
        @Synchronized
        @JvmStatic
        fun toSimpleDateFormat(format: String): SimpleDateFormat {
            var simpleDateFormat: SimpleDateFormat
            if (simpleDateFormats[format].also { simpleDateFormat = it!! } == null) {
                simpleDateFormats[format] = SimpleDateFormat(format).also { simpleDateFormat = it }
            }
            // SimpleDateFormat is not thread-safe
            // solution is return cloned object
            return simpleDateFormat.clone() as SimpleDateFormat
        }

        /**
         * format
         * @param date
         * @param format
         * @return
         */
        @Synchronized
        @JvmStatic
        fun format(date: Date?, format: String): String {
            return toSimpleDateFormat(format).format(date)
        }

        /**
         *
         * @param date
         * @return
         */
        @JvmStatic
        fun toCalendar(date: Date?): Calendar {
            val calendar = Calendar.getInstance()
            calendar.time = date
            return calendar
        }

        /**
         *
         * @param zonedDateTime
         * @return
         */
        @JvmStatic
        fun toCalendar(zonedDateTime: ZonedDateTime): Calendar {
            val calendar = Calendar.getInstance()
            calendar.clear()
            calendar[zonedDateTime.year, zonedDateTime.monthValue - 1, zonedDateTime.dayOfMonth, zonedDateTime.hour, zonedDateTime.minute] =
                zonedDateTime.second
            calendar[Calendar.MILLISECOND] = Math.round((zonedDateTime.nano / 1000000).toFloat())
            calendar.timeZone = TimeZone.getTimeZone(zonedDateTime.zone)
            return calendar
        }

        /**
         *
         * @param localDateTime
         * @return
         */
        @JvmStatic
        fun toCalendar(localDateTime: LocalDateTime): Calendar {
            val calendar = Calendar.getInstance()
            calendar.clear()
            calendar[localDateTime.year, localDateTime.monthValue - 1, localDateTime.dayOfMonth, localDateTime.hour, localDateTime.minute] =
                localDateTime.second
            calendar[Calendar.MILLISECOND] = localDateTime.nano / 1000000
            return calendar
        }

        /**
         *
         * @param localDate
         * @return
         */
        @JvmStatic
        fun toCalendar(localDate: LocalDate): Calendar {
            val calendar = Calendar.getInstance()
            calendar.clear()
            calendar[localDate.year, localDate.monthValue - 1] = localDate.dayOfMonth
            return calendar
        }

        /**
         *
         * @param localTime
         * @return
         */
        @JvmStatic
        fun toCalendar(localTime: LocalTime): Calendar {
            val calendar = Calendar.getInstance()
            calendar.clear()
            calendar[Calendar.HOUR] = localTime.hour
            calendar[Calendar.MINUTE] = localTime.minute
            calendar[Calendar.SECOND] = localTime.second
            calendar[Calendar.MILLISECOND] = localTime.nano / 1000000
            return calendar
        }
    }


}